spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/gson/GsonAutoConfiguration.java

```diff
@@ -16,28 +16,176 @@
 
 package org.springframework.boot.autoconfigure.gson;
 
+import com.google.gson.FieldNamingPolicy;
 import com.google.gson.Gson;
 
+import com.google.gson.GsonBuilder;
+import com.google.gson.LongSerializationPolicy;
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+
+import org.springframework.context.annotation.Primary;
+import org.springframework.core.Ordered;
+import org.springframework.util.CollectionUtils;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.List;
 
 /**
  * {@link EnableAutoConfiguration Auto-configuration} for Gson.
  *
  * @author David Liu
+ * @author Ivan Golovko
  * @since 1.2.0
  */
 @Configuration
 @ConditionalOnClass(Gson.class)
 public class GsonAutoConfiguration {
 
-	@Bean
-	@ConditionalOnMissingBean
-	public Gson gson() {
-		return new Gson();
-	}
+    @Configuration
+    @ConditionalOnClass({Gson.class})
+    static class GsonConfiguration {
+
+        @Bean
+        @Primary
+        @ConditionalOnMissingBean(Gson.class)
+        public Gson gson(GsonBuilder gsonBuilder) {
+            return gsonBuilder.create();
+        }
+    }
+
+    @Configuration
+    @ConditionalOnClass({Gson.class})
+    static class GsonBuilderConfiguration {
+
+        @Bean
+        @ConditionalOnClass({Gson.class})
+        public GsonBuilder gsonBuilder(
+                List<GsonBuilderCustomizer> customizers
+        ) {
+            final GsonBuilder gsonBuilder = new GsonBuilder();
+            customizers.forEach(c -> c.customize(gsonBuilder));
+            return gsonBuilder;
+        }
+
+    }
```

Comment:
There is no need for that dance. Rather than having two inner configuration class you can just do all of that in the body of the existing method. Is there a reason why you want to expose the `GsonBuilder`? Are you trying to mimic what we do for Jackson? (hence the primary). I think we do not need to go that far unless you have a good reason to.

Again, I'd prefer if this PR focuses on the configuration part only.